### PR TITLE
[Bug Fix] Fix several sparse optimizer bugs

### PR DESCRIPTION
**Pull request by classicsong**
## Description
Fix two bugs:
1. When NodeEmbedding is invoked multiple times in forward, grads and indices need to be collected together before doing grad update
2. Avoid using pytorch TCPStore to share information across training, use DGL shared memory instead.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/optim/pytorch/sparse_optim.py [hunk 1]
```diff
@@ -25,6 +25,35 @@ def __init__(self, params, lr):
...
+            else:
+                assert self._rank == emb.rank, \
+                    'MultiGPU rank for each embedding should be same.'
+                assert self._world_size == emb.world_size, \
+                    'MultiGPU world_size for each embedding should be same.'
+
+            emb_name = emb.name
+            if self._rank <= 0 and self._world_size > 1:
+                opt_meta = create_shared_mem_array(emb_name+'_opt_meta', \
+                    (self._world_size, self._world_size), th.int32).zero_()
```
Comment (zheng-da) [inline python/dgl/optim/pytorch/sparse_optim.py:46]: when `_rank` is < 0?

File: python/dgl/optim/pytorch/sparse_optim.py [hunk 2]
```diff
@@ -25,6 +25,35 @@ def __init__(self, params, lr):
...
+            if self._rank == 0:
+                if self._world_size > 1:
```
Comment (zheng-da) [inline python/dgl/optim/pytorch/sparse_optim.py:48]: how is this condition different from the one above (line 44)?

File: python/dgl/optim/pytorch/sparse_optim.py [hunk 3]
```diff
@@ -45,83 +74,101 @@ def step(self):
...
+                device = grad.device
+                idx_dtype = idx.dtype
+                grad_dtype = grad.dtype
+                grad_dim = grad.shape[1]
+
+                if self._world_size > 1:
+                    if emb_name not in self._shared_cache:
+                        self._shared_cache[emb_name] = {}
+
+                    for i in range(self._world_size):
```
Comment (zheng-da) [inline python/dgl/optim/pytorch/sparse_optim.py:95]: i think it's better to write comments to explain what this for loop does.

File: python/dgl/optim/pytorch/sparse_optim.py (hunk 1 above)
Comment (classicsong) [reply python/dgl/optim/pytorch/sparse_optim.py:46]: In either single gpu or multi gpu training. first process need to create the tensor.

Comment (classicsong) [reply python/dgl/optim/pytorch/sparse_optim.py:46]: In either single gpu or multi gpu training, the first process needs to create the tensor.

File: python/dgl/optim/pytorch/sparse_optim.py (hunk 2 above)
Comment (classicsong) [reply python/dgl/optim/pytorch/sparse_optim.py:48]: For single gpu training, rank is also 0

File: python/dgl/optim/pytorch/sparse_optim.py (hunk 1 above)
Comment (zheng-da) [reply python/dgl/optim/pytorch/sparse_optim.py:47]: I'm still very confused when rank < 0. As you said below, rank is 0 for single gpu. for multiple gpu training, i think the rank of the first gpu should also be 0. when it < 0?

File: python/dgl/optim/pytorch/sparse_optim.py (hunk 2 above)
Comment (zheng-da) [reply python/dgl/optim/pytorch/sparse_optim.py:49]: but then you have `if self._world_size > 1`. Is the condition the same as above?

File: python/dgl/optim/pytorch/sparse_optim.py [hunk 4]
```diff
@@ -36,92 +66,127 @@ def step(self):
...
+                        else:
+                            # currently nccl does not support Alltoallv operation
+                            # we need to use CPU shared memory to share gradient
+                            # across processes
+                            idx_i = idx_i.to(th.device('cpu'))
+                            grad_i = grad_i.to(th.device('cpu'))
+                            idx_shmem_name = 'idx_{}_{}_{}'.format(emb_name, self._rank, i)
+                            grad_shmem_name = 'grad_{}_{}_{}'.format(emb_name, self._rank, i)
+
+                            # Create shared memory to hold temporal index and gradient tensor for
```
Comment (zheng-da) [inline python/dgl/optim/pytorch/sparse_optim.py:127]: temporal? or temporary?

File: python/dgl/optim/pytorch/sparse_optim.py (hunk 1 above)
Comment (classicsong) [reply python/dgl/optim/pytorch/sparse_optim.py:47]: I c what you mean.

File: python/dgl/optim/pytorch/sparse_optim.py (hunk 4 above)
Comment (classicsong) [reply python/dgl/optim/pytorch/sparse_optim.py]: temporary

**Review by zheng-da — APPROVED** [review]

**Pull request merged**